### ISSUE TITLE
Re-OCR PDFs

### ISSUE THREAD

**Issue by finn-matti** [akosbalasko/yarle#82]
This might be out of scope, but it might be interesting: OCR PDFs (maybe through [OCRmyPDF](https://ocrmypdf.readthedocs.io/en/latest/)? Is there something more native to JS/TS?) that are included as attachments. Evernote describes their image recognition [here](https://evernote.com/blog/how-evernotes-image-recognition-works/):

>If you export a note containing a PDF that has been processed by the OCR system, there will be two nodes in the document: data and alternate-data . The data node contains a base–64 encoded version of the original PDF and the alternative-data represents the searchable version of the same PDF.

In [the DevonThink Forums](https://discourse.devontechnologies.com/t/how-to-batch-ocr-15-000-evernote-notes-many-of-which-contain-pdfs/59050/16?u=matti) one person from that company said that the OCR-Data is not available, but the above statement makes it sound it would be included.

**Comment by akosbalasko** [comment]
Hi @finn-matti !
Sorry for the late reply, I have played a bit with the idea you raised.
As I see there are two options around Evernote's recognitions: one is to make PDFs searchable, what you raised here, and another is to do OCR on pictures. In the first case, there is no information stored about the recognition in the Enex file, therefore, as you proposed, we could perform the recognition on our own (I mean as yarle of course). My problem with this solution is that we cannot guarantee that what we find is going to be identical to what is stored in EN. But we could give it a try, as I read somewhere that processing PDFs in EN works only on PDFs that allow selecting text within. So we may not need any OCR-magic, just extract the text from the PDF. 
The second case if we would like to process pictures OCRed by EN already, which result is stored in EN as <recognition>, with the found text and its position. So based on the coordinates yarle could restructure the text found by EN. What, IMHO, would be a really cool achievement. 
Then I converted a PDF to JPG, uploaded it as attachment in EN, exported it to Enex, and .... no recognition there. I think some magic within EN distinguishes attachments by file format and maybe by handwriting/typed (?) but this is just a tip, I'm not sure. 
In conclusion OCR is great, I'm not sure that it is required for extracting text from PDF attachments (for instance using this: https://www.npmjs.com/package/pdf-parse), and OCRing a picture, which means parsing recognition structure in enex, and reconstruct the text from it, looks a really promising and cool story. What do you think?

**Comment by finn-matti** [comment]
Hey @akosbalasko. Some notes:

>I read somewhere that processing PDFs in EN works only on PDFs that allow selecting text within. So we may not need any OCR-magic, just extract the text from the PDF.

I think it's actually the opposite. From the article I linked above:

>For a PDF to be eligible for OCR, it must meet certain requirements:
>1. It must contain a bitmap image
>2. It must not contain selectable text (or, at least, a minimal amount)

The second part is a little confusing to me. I think the goal is to have selectable and searchable text in the pdf - if not already present (and if it was already present). I would not know why to extract the text from pdfs or where to put it.

>The second case if we would like to process pictures OCRed by EN already, which result is stored in EN as , with the found text and its position. So based on the coordinates yarle could restructure the text found by EN.

Maybe we would have to make pdfs out of all those pictures to add a text-layer. I agree that it would be super-cool. And with recognition-data available to us, those pictures wouldn't need to be OCRed.

>Then I converted a PDF to JPG, uploaded it as attachment in EN, exported it to Enex, and .... no recognition there.

I tried something similar for testing something else and had OCR-recognition… in a pdf - so no recognition Data in the Export, again. (In this case: Screenshot of PDF with OCR -> made a pdf out of this screenshot -> inserted the screenshot-PDF into a Note and could search in the Screenshot-PDF after a while.) Maybe it just takes some time to get recognition data?

>In conclusion

To me two things would be cool to implement:

1. Make PDF-attachments searchable that aren't already - by using OCR
2. Make Pictures with recognition data searchable by converting them to PDF and adding a text-layer with the recognition data included within the note - no OCR is needed (maybe, if we could figure out how to do it)

**Comment by tokeriis** [comment]
Maybe pdf conversion could be achieved by [ocrmypdf](https://github.com/jbarlow83/OCRmyPDF)? 

It can be set to only OCR pdf files that have more than 0 characters.

**Comment by akosbalasko** [comment]
Hi @tokeriis !
As I see its a Python code, so it can be integrated into Yarle by a workaround, instead of I would like to have a nodejs lib to use.

**Comment by Mayeu** [comment]
Hello,

It seems that on some export the OCR data is available.

Here is such an example, I have a note which only contains [this comic](https://www.smbc-comics.com/comic/2012-09-02), when I export it in enex, I do have the OCR data at the end (it's pretty ugly, position is almost letter by letter):

<img width="933" alt="Screenshot 2022-01-01 at 23 29 44" src="https://user-images.githubusercontent.com/222375/147861442-2960a5fe-df29-4794-a773-647e5e40ef35.png">

Here is the [exported enex file](https://github.com/akosbalasko/yarle/files/7798311/My.Notes.enex.zip).

Note that I have done the export with [Evernote Legacy](https://help.evernote.com/hc/en-us/articles/360052560314-Install-an-older-version-of-Evernote) (the old v7 before the new "fancy" v10).
